src/xkernel.cpp

```diff
@@ -61,28 +64,57 @@ namespace xeus
                      interpreter_ptr interpreter,
                      history_manager_ptr history_manager,
                      server_builder builder)
-        : m_config(config)
-        , m_user_name(user_name)
-        , p_interpreter(std::move(interpreter))
-        , p_history_manager(std::move(history_manager))
-        , m_builder(builder)
+        : m_config(config),
+          m_user_name(user_name),
+          p_interpreter(std::move(interpreter)),
+          p_history_manager(std::move(history_manager)),
+          m_builder(builder)
     {
+        init();
+    }
+
+    xkernel::xkernel(const std::string& user_name,
+                     interpreter_ptr interpreter,
+                     history_manager_ptr history_manager,
+                     server_builder builder)
+        : m_user_name(user_name),
+          p_interpreter(std::move(interpreter)),
+          p_history_manager(std::move(history_manager)),
+          m_builder(builder)
+    {
+        init();
     }
 
     xkernel::~xkernel() {}
 
-    void xkernel::start()
+    void xkernel::init()
     {
         m_kernel_id = new_xguid();
         m_session_id = new_xguid();
 
+        if (m_config.m_key.size() == 0)
+        {
+            std::string ascii = "0123456789abcdefghijklmnopqrstuvwxyz";
+
+            std::random_device r;
+            std::default_random_engine generator(r());
+            std::uniform_int_distribution<std::size_t> distribution(0, 35);
+
+            m_config.m_key.resize(33);
+            for (auto& ch: m_config.m_key)
+            {
+                ch = ascii[distribution(generator)];
+            }
+            m_config.m_key[8] = '-';
+        }
+
         using authentication_ptr = xkernel_core::authentication_ptr;
         authentication_ptr auth = make_xauthentication(m_config.m_signature_scheme, m_config.m_key);
 
-        zmq::multipart_t start_msg;
-        build_start_msg(auth, m_kernel_id, m_user_name, m_session_id, start_msg);
+        build_start_msg(auth, m_kernel_id, m_user_name, m_session_id, m_start_msg);
 
         p_server = m_builder(m_context, m_config);
+        p_server->update_config(m_config);
```

Comment:
This is not required if the config ports are not empty.

Reply:
Yes